src/logic/safe/store/actions/createTransaction.ts

```diff
@@ -78,181 +75,253 @@ const navigateToTx = (safeAddress: string, txDetails: TransactionDetails) => {
+      promiEvent.once('error', reject)
+    })
+  }
 
-      const tx = isExecution ? getExecutionTransaction(txArgs) : getApprovalTransaction(safeInstance, safeTxHash)
-      const sendParams: PayableTx = {
-        from,
-        value: 0,
-        gas: ethParameters?.ethGasLimit,
-        [getGasParam()]: ethParameters?.ethGasPriceInGWei,
-        nonce: ethParameters?.ethNonce,
+  async submitTx(
+    state: AppReduxState,
+    confirmCallback?: ConfirmEventHandler,
+    errorCallback?: ErrorEventHandler,
+  ): Promise<void> {
+    // Off-chain signature
+    if (!this.isExecution) {
+      try {
+        const { hardwareWallet, smartContractWallet } = providerSelector(state)
+        const signature = await this.onlyConfirm(hardwareWallet, smartContractWallet)
+        this.onComplete(signature, confirmCallback)
+        return
+      } catch (err) {
+        logError(Errors._814, err.message)
       }
+    }
 
-      await tx
-        .send(sendParams)
-        .once('transactionHash', async (hash) => {
-          onUserConfirm?.(safeTxHash)
-
-          txHash = hash
-          dispatch(closeSnackbarAction({ key: beforeExecutionKey }))
-
-          try {
-            const txDetails = await saveTxToHistory({ ...txArgs, origin })
-
-            if (navigateToTransactionsTab) {
-              navigateToTx(safeAddress, txDetails)
-            }
-          } catch (err) {
-            logError(Errors._803, err.message)
-
-            // If we're just signing but not executing the tx, it's crucial that the request above succeeds
-            if (!isExecution) {
-              return
-            }
-          }
-
-          // store the pending transaction's nonce
-          isExecution && aboutToExecuteTx.setNonce(txArgs.nonce)
-
-          dispatch(fetchTransactions(chainId, safeAddress))
-        })
-        .then(async (receipt) => {
-          dispatch(fetchTransactions(chainId, safeAddress))
-          return receipt.transactionHash
-        })
+    // On-chain signature or execution
+    try {
+      await this.sendTx()
+      this.onComplete(undefined, confirmCallback)
     } catch (err) {
-      logError(Errors._803, err.message)
-      onError?.()
+      this.onError(err, errorCallback)
+    }
+  }
+
+  async prepare(dispatch: Dispatch, state: AppReduxState, txProps: RequiredTxProps): Promise<void> {
+    // Wallet connection
+    const ready = await onboardUser()
+    if (!ready) {
+      throw Error('No wallet connection')
+    }
 
-      dispatch(closeSnackbarAction({ key: beforeExecutionKey }))
+    // Selectors
+    const { account } = providerSelector(state)
+    this.from = account
 
-      if (isExecution && safeTxHash) {
-        dispatch(updateTransactionStatus({ safeTxHash, status: LocalTransactionStatus.PENDING_FAILED }))
+    this.safeVersion = currentSafeCurrentVersion(state)
+    const safeInstance = getGnosisSafeInstanceAt(txProps.safeAddress, this.safeVersion)
```

Comment:
It looks like this is used multiple times. Maybe we can move it to `this.safeInstance`? I'm not 100% sure though.